### PR TITLE
Add readonly dictionary wrapper that removes enumerator alloc

### PR DESCRIPTION
**Pull request by smoogipoo**
I find it pretty silly that this is still a thing even going into .NET 8. Hoping to use this for https://github.com/ppy/osu/pull/26863.

|                   Method |     Mean |    Error |   StdDev |    Gen0 | Allocated |
|------------------------- |---------:|---------:|---------:|--------:|----------:|
|               Dictionary | 35.05 us | 0.692 us | 1.035 us |       - |         - |
|     DictionaryAsReadOnly | 91.75 us | 1.833 us | 4.212 us | 22.9492 |   48000 B |
| DictionaryAsSlimReadOnly | 34.67 us | 0.690 us | 1.345 us |       - |         - |
|                     Keys | 28.73 us | 0.079 us | 0.074 us |       - |         - |
|           KeysAsReadOnly | 81.24 us | 0.101 us | 0.095 us | 19.0430 |   40000 B |
|       KeysAsSlimReadOnly | 28.75 us | 0.056 us | 0.052 us |       - |         - |
|                   Values | 28.50 us | 0.083 us | 0.074 us |       - |         - |
|         ValuesAsReadOnly | 81.41 us | 1.241 us | 0.969 us | 19.0430 |   40000 B |
|     ValuesAsSlimReadOnly | 28.58 us | 0.074 us | 0.058 us |       - |         - |

In contrast to `SlimReadOnlyListWrapper<T>`, the only thing gained by implementing any of the other interfaces is `CopyTo()`, so I've taken the liberty of not implementing every dictionary interface in existence here.

### CODE REVIEW

File: osu.Framework/Lists/SlimReadOnlyDictionaryWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        private readonly Dictionary<TKey, TValue> dict;
+
+        public SlimReadOnlyDictionaryWrapper(Dictionary<TKey, TValue> dict)
+        {
+            this.dict = dict;
+        }
+
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator()
+            => dict.GetEnumerator();
```
Comment (bdach) [inline osu.Framework/Lists/SlimReadOnlyDictionaryWrapper.cs:21]: This (and the non-generic `IEnumerator` one) still boxes, but I guess `SlimReadOnlyListWrapper` does the same so maybe fine.

File: osu.Framework.Benchmarks/BenchmarkSlimReadOnlyDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+            return sum;
+        }
+
+        [Benchmark]
+        public int DictionaryAsReadOnly()
+        {
+            int sum = 0;
+
+            ReadOnlyDictionary<int, int> readOnlyDictionary = new ReadOnlyDictionary<int, int>(dictionary);
```
Comment (bdach) [inline osu.Framework.Benchmarks/BenchmarkSlimReadOnlyDictionary.cs:43]: Maybe a bit pedantic but this alloc should be in `GlobalSetup()`.

File: osu.Framework.Benchmarks/BenchmarkSlimReadOnlyDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+            return sum;
+        }
+
+        [Benchmark]
+        public int KeysAsReadOnly()
+        {
+            int sum = 0;
+
+            ReadOnlyDictionary<int, int> readOnlyDictionary = new ReadOnlyDictionary<int, int>(dictionary);
```
Comment (bdach) [inline osu.Framework.Benchmarks/BenchmarkSlimReadOnlyDictionary.cs:87]: As above

File: osu.Framework/Lists/SlimReadOnlyDictionaryWrapper.cs (hunk 1 above)
Comment (smoogipoo) [reply osu.Framework/Lists/SlimReadOnlyDictionaryWrapper.cs:21]: This is fine, and is the way you're supposed to implement this. The only relevant case is `foreach ((SlimReadOnlyDictionary<>)dict)` - if you go via the interface then you're SOL no matter what.

File: osu.Framework/Extensions/ListExtensions/ListExtensions.cs [hunk 4]
```diff
@@ -17,5 +17,17 @@ public static class ListExtensions
...
+
+        /// <summary>
+        /// Creates a new read-only view into a <see cref="Dictionary{TKey, TValue}"/>.
+        /// </summary>
+        /// <remarks>Enumeration does not allocate the enumerator.</remarks>
+        /// <param name="dictionary">The dictionary to create the view of.</param>
+        /// <typeparam name="TKey">The type of keys in the dictionary.</typeparam>
+        /// <typeparam name="TValue">The type of values in the dictionary.</typeparam>
+        /// <returns>The read-only view.</returns>
+        public static SlimReadOnlyDictionaryWrapper<TKey, TValue> AsSlimReadOnly<TKey, TValue>(this Dictionary<TKey, TValue> dictionary)
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Framework/Extensions/ListExtensions/ListExtensions.cs:29]: I could split it out into a `DictionaryExtensions`, but it felt like too much. If you'd feel better with that then let me know and I'll do it.

**Review by bdach — APPROVED** [review]

**Pull request merged**